Resampling and Translation Bug

cmarshak opened the pull request:
Addresses #32.

- [x] Include tests for updates including integration test specified within the issue ticket
- [x] Clarify readme for how transformations are performed (including the order)
- [x] Returns the API to the original form (i.e. returns `(X, p)` where `X` is array and `p` is profile

cmarshak commented:
I am going to write one more test for the geoid reader.

cmarshak commented:
@asjohnston-asf - when it makes sense, can you help me understand why one of my tests is failing here (i.e. `tests/test_rio_tools.py`? I thought it might be because data was correctly uploaded to the repo, then I successfully tried the tests by cloning it on mac and then on a linux server. Honestly, not totally sure how to debug. It looks like the first array is all 0's but it shouldn't be...

The data is here: https://github.com/ACCESS-Cloud-Based-InSAR/dem-stitcher/tree/reproject-and-test/tests/data/rio_tools/update_resolution

Here is one failed action. https://github.com/ACCESS-Cloud-Based-InSAR/dem-stitcher/runs/6564288126?check_suite_focus=true

I temporarily marked it as an integration test so it's ignored.

Maybe it's an action issue?

cmarshak commented:
This is way too large of a pull request for a time-strapped team to review carefully (40 files changed) and so I understand there is going to be some element of trust. I am putting some of the higher level elements of the review.

## Integration with DockerizedTopsApp

Also, as apart of this pull request - is the related one [here](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/pull/66) in DockerizedTopsApp. In the pull request, note we pull this branch in the environment [file](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/dem-updates/environment.yml#L44). I ran the code on 4 different areas (their selection was motivated by previous issue tickets) and the resulting GUNWs can be downloaded there. Ultimatley, the 90 meter GUNW doesn't show much of a difference using this new dem.

Most important for @sssangha and @gracebato to take a look at the datasets and just confirm they look reasonable and I am not missing something. The new Dem-stitcher does not extend the origin of DEM tiles if they are not available (this is a feature, not a bug). The resulting GUNW over La Palma is truncated - but its georeferencing is fine as can be inspected.

## Documentation about "stitching" and "resampling"

Within this pull request and now the readme, I tried to explain how the tiles are stitched and transformed.

I also included two [notebooks](https://github.com/ACCESS-Cloud-Based-InSAR/dem-stitcher/tree/reproject-and-test/notebooks/analysis_and_comparison) to demonstrate that (1) if no geoid is removed the tiles are subset but not resampled and (2) to compare ISCE2 dem.py to our stitcher.

@jhkennedy and @sssangha - I would like you to provide feedback.

## API Changes

1. We add `dst_resolution` so that we can specify a target square resolution or `(x_res, y_res)` solution. We make sure that we do not adjust the origin as is frequently done in `rasterio`.
2. We return the API to return the tuple of `(X, p)`. This will greatly support ACCESS's "virtual data processing" (processing in memory) since we can pass the arrays to subsequent workflows directly.
3. I added a docstring to the main `stitch` function.

## Tests

There are ton of tests that make sure that we are correctly stitching the tiles together. Here are some high level descriptions:

1. In [test_window.py](https://github.com/ACCESS-Cloud-Based-InSAR/dem-stitcher/blob/reproject-and-test/tests/test_window.py), we want to make sure when we grab an extent within an image we do not do *any* resampling. There are additional tests associated if the extent is outside the image's extent.
4. In [test_stitcher.py](https://github.com/ACCESS-Cloud-Based-InSAR/dem-stitcher/blob/reproject-and-test/tests/test_stitcher.py), we are making sure that when we merge there again is no resampling. One of the [tests](https://github.com/ACCESS-Cloud-Based-InSAR/dem-stitcher/blob/reproject-and-test/tests/test_stitcher.py#L146) is distilling [this](https://github.com/ACCESS-Cloud-Based-InSAR/dem-stitcher/blob/reproject-and-test/notebooks/analysis_and_comparison/0_Alignment_with_original_DEM_tiles.ipynb) notebook from above and making sure when we compare the stitched DEM to the original tiles and do not remove geoid (and do not alter the origin) we are not modifying the original tile data at all.
5. [test_geoid.py](https://github.com/ACCESS-Cloud-Based-InSAR/dem-stitcher/blob/reproject-and-test/tests/test_geoid.py) does some combination of the first two tests in the context of reading geoids. It also illustrates what a reasonable buffer which depends on the resolution of the geoid relative to the DEM raster that it is being removed from.
6. There is also [test_rio_tools](https://github.com/ACCESS-Cloud-Based-InSAR/dem-stitcher/blob/reproject-and-test/tests/test_rio_tools.py) which makes sure that when we upsample the origin does not change.

There is also an integration test (which is ignored by github actions) which does reach onto the servers to make a stitched DEM, but makes sure the main `stitch` function goes until completion.

## Packaging

Unfortunatley, there were packaging issues (see [1] and [2]) that cropped up due to the changes related to `gdal` and `rasterio` since I worked with them last that made this pull request harder. It was essential that the DEM stitcher still worked with DockerizedTopsApp which uses `conda-forge` packages specifically its `gdal`. Long story short, we found that `gdal` had to be version `<3.4`. We also discovered that the PyPI `rasterio` is not installing correctly for python 3.10+ as in [2] (we have an environment that installs a beta release but it is not tested here).

These issues were first exposed after writing all the tests and passing in an environment with an older python worked fine but they would fail on GitHub actions. After some help from @jhkennedy, it was found that all resampling was defaulting to `nearest` in certain environments and therefore the above `test_rio_tools.py` was failing; this was no good to say the least (wanted our geoid to be resampled smoothly to the DEM tiles!). After some investigations, we have settled on the current solution. I hope that @jhkennedy and @asjohnston-asf might help make this cleaner in the future. Currently, I have two environment files (one for PyPi and one for conda) and they are both tested against the new suite - the key is `gdal` (and maybe `proj`).

[1] https://github.com/conda-forge/rasterio-feedstock/issues/243 
[2] https://github.com/rasterio/rasterio/issues/2473


## Code Organization

Please provide feedback if there is time!

cmarshak commented:
To illustrate that the outputs preserve the georeferencing of the original DEM tiles, here is an example output over a small .1 degree area overlapping 4 tiles in Central CA.

Consider the following:

```
from dem_stitcher.stitcher import stitch_dem, get_dem_tiles
import rasterio

dst_area_or_point = 'Point'
dst_ellipsoidal_height = False
dem_name = 'glo_30'

# Central coast of California
# xmin, ymin, xmax, ymax
bounds = [-120.05, 34.95, -119.95, 35.05]

X, p = stitch_dem(bounds,
                  dem_name=dem_name,
                  dst_ellipsoidal_height=dst_ellipsoidal_height,
                  dst_area_or_point=dst_area_or_point)

height_type = 'ellipsoidal' if dst_ellipsoidal_height else 'geoid'

with rasterio.open(f'{dem_name}_{height_type}_{dst_area_or_point}_CA.tif', 'w', **p) as ds:
    ds.write(X, 1)
    ds.update_tags(AREA_OR_POINT=dst_area_or_point) 
```

The output generates [this](https://gunw-development-testing.s3.us-west-2.amazonaws.com/dem-updates-May-2022/glo_30_geoid_Point_CA.tif).


The source tiles are here
```
https://copernicus-dem-30m.s3.amazonaws.com/Copernicus_DSM_COG_10_N34_00_W121_00_DEM/Copernicus_DSM_COG_10_N34_00_W121_00_DEM.tif
https://copernicus-dem-30m.s3.amazonaws.com/Copernicus_DSM_COG_10_N34_00_W120_00_DEM/Copernicus_DSM_COG_10_N34_00_W120_00_DEM.tif
https://copernicus-dem-30m.s3.amazonaws.com/Copernicus_DSM_COG_10_N35_00_W121_00_DEM/Copernicus_DSM_COG_10_N35_00_W121_00_DEM.tif
https://copernicus-dem-30m.s3.amazonaws.com/Copernicus_DSM_COG_10_N35_00_W120_00_DEM/Copernicus_DSM_COG_10_N35_00_W120_00_DEM.tif
```

If you look at them in QGIS, you should see no shift across any of the tiles.

jhkennedy commented on environment.yml:
```diff
@@ -4,28 +4,30 @@ channels:
+    - geopandas>=0.5.1
+    - numpy>=1.17.2
+    - pandas>=0.25.1
+    - rasterio>=1.0.28
```
Any reason this isn't just `rasterio>=1.2` like in `environments/environment-isce.yml` (and also the inverse; why the constraint there?)

cmarshak replied:
Yea - the only problematic version is `gdal` currently. Hopefully that resolves in the not-to-distant future.

jhkennedy reviewed (commented):
After a first pass, I have quite a few questions around the environments. I *also* have a lot of suggestions/changes for the environment files, that I'll hold off on for now until some of those questions are answered, or at least discussed. Afterwards, I think I'll just open a PR to this PR with those suggestions as there a quite a few of them.

cmarshak commented on tests/test_stitcher.py:
```diff
@@ -1,20 +1,191 @@
+src_tags = ['Area',
+            'Area',
+            'Point',
+            'Point'
+            ]
+
+dst_tags = ['Area',
+            'Point',
+            'Point',
+            'Area']
```
Done

cmarshak commented on README.md:
```diff
@@ -74,27 +83,33 @@ Out[1]: ['srtm_v3', 'nasadem', 'glo_30', '3dep', 'ned1']
+1. Create an GitHub issue ticket desrcribing what changes you would like to see or to report a bug
+2. We will work on solving this issue (hopefully with you)
 
-Create an issue ticket.
```
I thought I saw this in an ASF repo. I was mistaken. Reverting.

cmarshak commented on dem_stitcher/dem_readers.py:
```diff
@@ -23,8 +23,14 @@ def read_glo(url: str) -> Union[rasterio.DatasetReader, None]:
+        data = io.BytesIO(resp.content)
+    # local file
+    else:
+        data = url
```
Sounds good.

cmarshak commented:
@jhkennedy 

- i removed all but `environment.yml`
- updated the readme to describe environment installation *and* that various environment setups is beyond the scope of this repository

I hope at some point we can add python 3.10 using conda - but won't worry about it now. Also, we can add conda install instructions once we get it onto conda-forge.

cmarshak commented on environment.yml:
```diff
@@ -4,28 +4,30 @@ channels:
- - setuptools
- - setuptools_scm
- - wheel
+ - pip:
```
I removed pypi dependencies and am focusing on `conda` with `conda-forge`. Good to know about that requirements have this control flow related to python version.

sssangha reviewed (approved):
Looks good, and with thorough documentation!

jhkennedy commented on setup.py:
```diff
@@ -24,23 +24,18 @@
         'geopandas',
         'importlib_metadata',  # drop when py>=3.8
         'numpy',
-        'pyproj',
```
You're calling pyproj in `rio_window.py`, so you should keep this one in `setup.py`

jhkennedy commented on setup.py:
```diff
@@ -24,23 +24,18 @@
-        'pyproj',
         'rasterio',
         'requests',
-        'shapely',
```
same here, you're using shapely in a couple places.

jhkennedy commented on setup.py:
```diff
@@ -24,23 +24,18 @@
 
     install_requires=[
         'affine',
         'boto3',
```
Since you're not importing/using this directly, you could add a comment string to say `# for rasterio vsis3 support`

jhkennedy reviewed (approved):
Overall looks good! Suggested some minor tweaks that aren't blocking

jhkennedy commented on environment.yml:
```diff
@@ -4,28 +4,26 @@ channels:
  - setuptools_scm
- - wheel
+ - shapely
+ - tqdm
```
```suggestion
 - tqdm
 
```

cmarshak replied:
done

cmarshak commented on environment.yml:
```diff
@@ -4,28 +4,26 @@ channels:
- - fiona
- - gdal
+ - dateparser
+ - Fiona
```
done

cmarshak commented on environment.yml:
```diff
@@ -4,28 +4,26 @@ channels:
  - setuptools
  - setuptools_scm
- - wheel
+ - Shapely
```
done
